orttraining/orttraining/core/framework/module_gradient_graph_builder.cc

```diff
@@ -77,8 +77,13 @@ Status ModuleGradientGraphBuilder::Build(const std::vector<std::vector<int64_t>>
     SetConcreteInputShapes(*input_shapes_ptr);
   }
 
-  // Build the gradient graph.
-  ORT_RETURN_IF_ERROR(BuildGradientGraph());
+  // Optimize the inference graph, and if needed, build the gradient graph.
+  ORT_RETURN_IF_ERROR(OptimizeInferenceGraphAndBuildGradientGraph());
+  if (!config_.build_gradient_graph) {
+    // This graph will be used only for inferencing, so stop right here.
+    // The gradient graph has not been built for this graph
+    return Status::OK();
+  }
```

Comment:
Sorry I wasn't clear on this! I was thinking something more like the following (to avoid the double-check):

```suggestion
  std::unordered_set<std::string> x_node_arg_names;
  ORT_RETURN_IF_ERROR(OptimizeInferenceGraph(x_node_arg_names));
  if (!config_.build_gradient_graph) {
    // This graph will be used only for inferencing, so stop right here.
    // No need to build a gradient graph.
    return Status::OK();
  }
  ORT_RETURN_IF_ERROR(BuildGradientGraph(x_node_arg_names));
```

...and then you don't need the `OptimizeInferenceGraphAndBuildGradientGraph()` method.

Reply:
Yeah, I understood the intent in the comment before. I did not want to add more implementation logic in this function since it felt more like a higher level function that delegated the tasks to other functions.
However, to avoid the double checks, I will make this change.

Reply:
Got it and thanks. Yeah, the `x_node_arg_names` being passed around at the top level is a bit of a wart here. There's probably a more invasive modification that could avoid it (but probably not worth doing in this PR!).